### PR TITLE
feat(accountsdb): improved metrics

### PR DESCRIPTION
**Pull request by Sobeston**
https://github.com/orgs/Syndica/projects/2?pane=issue&itemId=68540818

Dashboard preview:
image

### CODE REVIEW

File: src/accountsdb/db.zig [hunk 1]
```diff
@@ -1447,7 +1509,7 @@ pub const AccountsDB = struct {
             const account_file, var account_file_lg = account_file_rw.readWithLock();
             defer account_file_lg.unlock();
 
-            // self.logger.debugf("cleaning slot: {}...", .{account_file.slot});
```
Comment (Sobeston) [inline src/accountsdb/db.zig:1450]: Was this commented out for a reason?

File: src/fuzz.zig [hunk 2]
```diff
@@ -1,11 +1,14 @@
 const std = @import("std");
 const lib = @import("./lib.zig");
+const config = @import("./cmd/config.zig");
```
Comment (Sobeston) [inline src/fuzz.zig:3]: Or should we configure the metrics port another way?

File: src/accountsdb/db.zig (hunk 1 above)
Comment (0xNineteen) [reply src/accountsdb/db.zig:1450]: yeah it was too much logs

File: src/accountsdb/db.zig [hunk 3]
```diff
@@ -1313,6 +1362,8 @@ pub const AccountsDB = struct {
     /// as the data field ([]u8) for each account.
     /// Returns the unclean file id.
     pub fn flushSlot(self: *Self, slot: Slot) !FileId {
+        var timer = std.time.Timer.start() catch @panic("Timer unsupported");
```
Comment (0xNineteen) [inline src/accountsdb/db.zig:1365]: should use `sig.time.Timer` here with a `try` ... - maybe we can move the `catch @panic` into the sig.time.Timer struct to remove the try in another pr

File: src/accountsdb/db.zig [hunk 4]
```diff
@@ -1548,8 +1610,19 @@ pub const AccountsDB = struct {
                 }
             }
             references_to_delete.clearRetainingCapacity();
+            self.stats.clean_references_deleted.set(references_to_delete.items.len);
+            self.logger.debugf(
+                "cleaned slot {} -  old_state: {}, zero_lamports: {}",
```
Comment (0xNineteen) [inline src/accountsdb/db.zig:1615]: extra space here

File: src/accountsdb/db.zig [hunk 5]
```diff
@@ -1548,8 +1610,19 @@ pub const AccountsDB = struct {
...
+                .{ account_file.slot, num_old_states, num_zero_lamports },
+            );
         }
 
+        self.stats.clean_files_queued_deletion.set(delete_account_files.count());
+        self.stats.clean_files_queued_shrink.set(delete_account_files.count());
+        self.stats.clean_slot_old_state.set(num_old_states);
+        self.stats.clean_slot_zero_lamports.set(num_zero_lamports);
+
+        self.logger.debugf("cleanAccountFiles time elapsed: {}ns", .{timer.read()});
```
Comment (0xNineteen) [inline src/accountsdb/db.zig:1625]: should probably move elapsed times into a bucket -- see how we handle it in gossip service `processMessages`

File: src/fuzz.zig (hunk 2 above)
Comment (0xNineteen) [reply src/fuzz.zig:3]: i think this is ok for now

File: src/fuzz.zig [hunk 6]
```diff
@@ -19,6 +22,13 @@ pub const FuzzFilter = enum {
     gossip_table,
 };
 
+fn spawnMetrics(gpa_allocator: std.mem.Allocator) !std.Thread {
+    const metrics_port: u16 = config.current.metrics_port;
+    logger.default_logger.infof("metrics port: {d}", .{metrics_port});
+    const registry = globalRegistry();
+    return std.Thread.spawn(.{}, servePrometheus, .{ gpa_allocator, registry, metrics_port });
+}
```
Comment (0xNineteen) [inline src/fuzz.zig:30]: can we move this into the prometheus file and call it from both here and `cmd.zig` (shouldnt need `spawnMetrics`) - seems like code duplication rn

File: src/accountsdb/db.zig [hunk 7]
```diff
@@ -1313,6 +1362,8 @@ pub const AccountsDB = struct {
     /// as the data field ([]u8) for each account.
     /// Returns the unclean file id.
     pub fn flushSlot(self: *Self, slot: Slot) !FileId {
+        var timer = std.time.Timer.start() catch @panic("Timer unsupported");
+
```
Comment (0xNineteen) [inline src/accountsdb/db.zig:1366]: outside of scope for a comment but i think with these metric changes we can delete the result returns on clean and shrink in the `runManagerLoop` and just return void instead (eg, change from `const clean_result = try self.cleanAccountFiles(` to `try self.cleanAccountFiles(` and delete `self.logger.debugf("clean_result: {any}", .{clean_result});` for both clean and shrink)

File: src/accountsdb/fuzz.zig [hunk 8]
```diff
@@ -39,6 +39,8 @@ pub const TrackedAccount = struct {
 };
 
 pub fn run(seed: u64, args: *std.process.ArgIterator) !void {
+    _ = try sig.accounts_db.db.AccountsDBStats.init();
```
Comment (0xNineteen) [inline src/accountsdb/fuzz.zig:42]: why do we need this? this is included in AccountsDB.init

**Review by 0xNineteen — CHANGES_REQUESTED** [review]

File: src/accountsdb/db.zig (hunk 7 above)
Comment (Sobeston) [reply src/accountsdb/db.zig:1378]: I'm thinking to make this change after removing the error union from Timer.start(). Happy to "resolve" this for now and pick it up later?

**Review by yewman — APPROVED** [review]
Looks good 👍

**Review by 0xNineteen — APPROVED** [review]
:shipit:

**Pull request merged**